2024/12/solve-1.ys

```diff
@@ -1,100 +1,72 @@
 #!/usr/bin/env ys-0
 
-defn main(& args):
-  terrain =: init(args.0 || "input")
-  regions =: terrain:regions
-  println:
-    total-price: terrain regions
-
-defn init(file="input"):
-  file:slurp:lines
+defn main(input='input'): !:say
+  terrain =: input:slurp:lines
+  regions =: regions(terrain)
+  total-price: terrain regions
 
 defn total-price(terrain regions):
-  sum:
-    map _ regions:
-      fn(r): price(terrain r)
-
-defn price(terrain region):
-  =>: region:len * perimeter(terrain region)
+  defn price(region):
+    region.# * perimeter(terrain region)
+  sum: regions.map(price)
 
 defn perimeter(terrain region):
-  sum:
-    map _ region:seq:
-      fn(yx):
-        len:
-          grep _ yx:nesw:
-            fn(it): region.get(it):not
-
-defn regions(terrain):
-  first:
-    reduce _ [[] hash-set()] terrain:coordinates:
-      fn([regions visited] yx):
-        if visited.get(yx):
-          vector: regions visited
-          else:
-            region =: terrain.region(terrain.at(yx) vector(yx))
-            vector: regions.conj(region) visited.union(region)
+  region.map(
+    \(nesw(_).remove(\(region.get(_))).#)
+  ):sum
+
+dirs =::
+- [1, 0]
+- [-1, 0]
+- [0, 1]
+- [0, -1]
+
+defn nesw([y x]):
+  for [d0 d1] dirs:
```

Comment:
`for x y:` is short for `for [x y]:`.
Same with
`if a > b:` short for `if (a > b):`
See https://github.com/yaml/yamlscript/blob/main/core/src/yamlscript/transformers.clj#L159-L173
for forms that auto-paren or auto-brace.